File: src/ru/henridellal/dialer/DialerPreferencesActivity.java
package ru.henridellal.dialer;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceActivity;
import android.preference.PreferenceManager;

public class DialerPreferencesActivity extends PreferenceActivity implements SharedPreferences.OnSharedPreferenceChangeListener {
	
	private boolean restartTriggered;
	private SharedPreferences preferences;
	
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		addPreferencesFromResource(R.xml.preferences);
		ListPreference themePreference = (ListPreference) findPreference("theme");
		themePreference.setSummary(themePreference.getEntry());
		preferences = PreferenceManager.getDefaultSharedPreferences(this);
	}
	
	@Override
	protected void onResume() {
		super.onResume();
		preferences.registerOnSharedPreferenceChangeListener(this);
	}
	
	@Override
	protected void onPause() {
		super.onPause();
		preferences.unregisterOnSharedPreferenceChangeListener(this);
	}
	
	@Override
	public void onBackPressed() {
		if (restartTriggered) {
			startActivity(new Intent(this, DialerActivity.class));
			finishAffinity();
		} else {
			super.onBackPressed();
		}
	}
	
	@Override
	public void onSharedPreferenceChanged(SharedPreferences prefs, String key) {
		Preference pref = findPreference(key);
		if (pref instanceof ListPreference) {
			ListPreference listPreference = (ListPreference) pref;
			listPreference.setSummary(listPreference.getEntry());
		}
	
		if ("theme".equals(key)
			|| "t9_locale".equals(key)) {
			restartTriggered = true;
		}
	}
}

File: app/src/main/java/ru/henridellal/dialer/DialerPreferencesActivity.kt
package ru.henridellal.dialer

import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.preference.ListPreference
import android.preference.PreferenceActivity
import android.preference.PreferenceManager

class DialerPreferencesActivity : PreferenceActivity(), OnSharedPreferenceChangeListener {
    private var restartTriggered = false
    private var preferences: SharedPreferences? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        addPreferencesFromResource(R.xml.preferences)
        val themePreference = findPreference("theme") as ListPreference
        themePreference.summary = themePreference.entry
        preferences = PreferenceManager.getDefaultSharedPreferences(this)
    }

    override fun onResume() {
        super.onResume()
        preferences!!.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        preferences!!.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onBackPressed() {
        if (restartTriggered) {
            startActivity(Intent(this, DialerActivity::class.java))
            finishAffinity()
        } else {
            super.onBackPressed()
        }
    }

    override fun onSharedPreferenceChanged(prefs: SharedPreferences, key: String) {
        val pref = findPreference(key)
        if (pref is ListPreference) {
            pref.summary = pref.entry
        }
        if ("theme" == key || "t9_locale" == key) {
            restartTriggered = true
        }
    }
}